src/index.ts

```diff
@@ -110,37 +80,86 @@ export function lambdaHandler<TContext extends BaseContext>(
-  parsedBody: string | object | undefined,
-  event: APIGatewayProxyEvent | APIGatewayProxyEventV2,
-): HTTPGraphQLRequest {
-  const headers = new Map<string, string>();
-  for (const [key, value] of Object.entries(event.headers)) {
-    if (value !== undefined) {
-      // Node/Express headers can be an array or a single value. We join
-      // multi-valued headers with `, ` just like the Fetch API's `Headers`
-      // does. We assume that keys are already lower-cased (as per the Node
-      // docs on IncomingMessage.headers) and so we don't bother to lower-case
-      // them or combine across multiple keys that would lower-case to the
-      // same value.
-      headers.set(key, Array.isArray(value) ? value.join(', ') : value);
-    }
+function normalizeGatewayEvent(event: GatewayEvent): HTTPGraphQLRequest {
+  if (isV1Event(event)) {
+    return normalizeV1Event(event);
+  } else if (isV2Event(event)) {
+    return normalizeV2Event(event);
+  } else {
+    throw Error('Unknown event type');
   }
+}
+
+function isV1Event(event: GatewayEvent): event is APIGatewayProxyEvent {
+  return !('version' in event);
+}
+
+function isV2Event(event: GatewayEvent): event is APIGatewayProxyEventV2 {
+  return 'version' in event && event.version === '2.0';
+}
+
+function normalizeV1Event(event: APIGatewayProxyEvent): HTTPGraphQLRequest {
+  const headers = normalizeHeaders(event.headers);
+  const body = parseBody(event.body, headers.get('content-type'));
+  const reconstructedSearchString = Object.entries({
+    ...event.queryStringParameters,
+    ...event.multiValueQueryStringParameters,
+  }).reduce((queryString, next, i) => {
+    const delimiter = i === 0 ? '?' : '&';
+    const addition = Array.isArray(next[1])
+      ? next[1].map((v) => `${next[0]}=${v}`).join('&')
+      : `${next[0]}=${next[1]}`;
+    return `${queryString}${delimiter}${addition}`;
+  }, '');
+
+  return {
+    method: event.httpMethod,
+    headers,
+    search: reconstructedSearchString,
+    body,
+  };
+}
+
+function normalizeV2Event(event: APIGatewayProxyEventV2): HTTPGraphQLRequest {
+  const headers = normalizeHeaders(event.headers);
+  return {
+    method: event.requestContext.http.method,
+    headers,
+    search: event.rawQueryString,
+    body: parseBody(event.body, headers.get('content-type')),
+  };
+}
 
-  if ((<APIGatewayProxyEventV2>event).requestContext.http !== undefined) {
-    return {
-      method: (<APIGatewayProxyEventV2>event).requestContext.http.method,
-      headers,
-      search: (<APIGatewayProxyEventV2>event).rawQueryString,
-      body: parsedBody,
-    };
+function parseBody(
+  body: string | null | undefined,
+  contentType: string | undefined,
+): object | string {
+  if (!body) {
+    return '';
+  } else if (contentType === 'application/json') {
+    return JSON.parse(body);
+  } else if (contentType === 'text/plain') {
+    return body;
   } else {
-    return {} as HTTPGraphQLRequest;
+    return '';
+  }
```

Comment:
Similar to above this could be collapsed due to the `return` short circuit.

```
  if (body && contentType === 'application/json') {
    return JSON.parse(body);
  }
  if (body && contentType === 'text/plain') {
    return body;
  }
  return '';
```

Reply:
Agree with collapsing this. Any preference on nesting within an `if (body) { ... }` vs. what you've proposed?

Reply:
I agree that'd be better!